core/schema/util.go

```diff
@@ -73,3 +75,121 @@ var AllVersion = core.AllVersion
 
 type BeforeVersion = core.BeforeVersion
 type AfterVersion = core.AfterVersion
+
+// Extract the gitignore patterns from its file content and return them as relative paths
+// based on the given parent directory.
+// The parent directory is required to match the gitignore patterns relative to
+// the context directory since the given content may be a .gitignore file from a subdirectory
+// of that context.
+//
+// Example:
+//   - contextDir = "/my-project/foo"
+//   - .gitignoreDir = "/my-project/foo/bar/.gitignore"
+//   - parentDir = "/my-project/foo/bar" -> so **/node_modules becomes my-project/foo/bar/**/node_modules
+//
+// Pattern additional formatting is based on https://git-scm.com/docs/gitignore so it can
+// be correctly applied with further dagger include/exclude filter patterns.
+//
+// Here are the rules of that filter/formatting:
+//
+//   - We ignore empty lines and comments (starting with `#`) (`\#` isn't ignored).
+//
+//   - If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative.
+//     Otherwise, the pattern is recursive.
+//     If pattern is already starting with **, not change needed.
+//     If the pattern starts with *, it's not recursive but only matches the directory itself.
+//
+//   - If a pattern is negative exclusion (starts with `!`) or targets directory only
+//     (ends with `/`), we treat is as a regular path then readd the exclusion to make
+//     sure the recusive pattern is applied if needed.
```

Comment:
Can you give examples of these?

Reply:
Done in [365929e](https://github.com/dagger/dagger/pull/10736/commits/365929ec62617b929b10a62820e26ec77e990f05)